src/wallet/wallet2.cpp

```diff
@@ -7917,6 +7917,198 @@ bool wallet2::check_tx_proof(const crypto::hash &txid, const cryptonote::account
+  {
+    const transfer_details &td = m_transfers[idx];
+    proofs.resize(proofs.size() + 1);
+    auto& proof = proofs.back();
+    proof.txid = td.m_txid;
+    proof.key = td.get_public_key();
+    proof.ki = td.m_key_image;
+    proof.amount = td.amount();
+
+    std::string prefix_data((const char*)&proof.txid, sizeof(crypto::hash));
+    prefix_data += std::string((const char*)&proof.key, sizeof(crypto::public_key));
+    prefix_data += message;
+    crypto::hash prefix_hash;
+    crypto::cn_fast_hash(prefix_data.data(), prefix_data.size(), prefix_hash);
+
+    // generate ephemeral secret key
+    const crypto::public_key tx_pub_key = get_tx_pub_key_from_extra(td.m_tx, td.m_pk_index);
+    THROW_WALLET_EXCEPTION_IF(tx_pub_key == crypto::null_pkey, error::wallet_internal_error, "The tx public key isn't found");
+    const std::vector<crypto::public_key> additional_tx_pub_keys = get_additional_tx_pub_keys_from_extra(td.m_tx);
+    crypto::key_image ki;
+    cryptonote::keypair ephemeral;
+    const bool r = cryptonote::generate_key_image_helper(m_account.get_keys(), m_subaddresses, td.get_public_key(), tx_pub_key,  additional_tx_pub_keys, td.m_internal_output_index, ephemeral, ki);
+    THROW_WALLET_EXCEPTION_IF(!r, error::wallet_internal_error, "Failed to generate key image");
+    THROW_WALLET_EXCEPTION_IF(ephemeral.pub != td.get_public_key(), error::wallet_internal_error, "Derived public key doesn't agree with the stored one");
+
+    if (td.m_tx.version == 1)
+    {
+      const std::vector<const crypto::public_key*> pubs = { &ephemeral.pub };
+      crypto::generate_ring_signature(prefix_hash, td.m_key_image, &pubs[0], 1, ephemeral.sec, 0, &proof.nonrctSig);
+    }
+    else
+    {
+      rct::ctkeyV pubs(1);
+      pubs[0].dest = rct::pk2rct(ephemeral.pub);
+      pubs[0].mask = rct::commit(td.amount(), td.m_mask);
+      rct::ctkey inSk;
+      inSk.dest = rct::sk2rct(ephemeral.sec);
+      inSk.mask = td.m_mask;
+      const rct::key a = rct::zero();
+      const rct::key C = rct::scalarmultH(rct::d2h(td.amount()));
+      proof.rctSig = rct::proveRctMGSimple(rct::hash2rct(prefix_hash), pubs, inSk, a, C, nullptr, nullptr, 0);
+    }
+  }
+
+  std::ostringstream oss;
+  boost::archive::portable_binary_oarchive ar(oss);
+  ar << proofs;
+
+  std::string sig_str = std::string("ReserveProofV1");
+  sig_str += tools::base58::encode(oss.str());
+  return sig_str;
+}
+
+bool wallet2::check_reserve_proof(const std::string &message, const std::string &sig_str, uint64_t &total, uint64_t &spent)
+{
+  const std::string header = "ReserveProofV1";
+  const size_t header_len = header.size();
+  THROW_WALLET_EXCEPTION_IF(sig_str.size() < header_len || sig_str.substr(0, header_len) != header, error::wallet_internal_error,
+    "Signature header check error");
+
+  std::string sig_decoded;
+  THROW_WALLET_EXCEPTION_IF(!tools::base58::decode(sig_str.substr(header_len), sig_decoded), error::wallet_internal_error,
```

Comment:
This would have to become `sig_str.substr(std::strlen(header)))` obviously. Or do some sizeof tricks.